lib/Quantization/Base/Base.cpp

```diff
@@ -566,28 +566,63 @@ TensorQuantizationParams
 specializeBiasQuantizationParams(const TensorQuantizationParams &biasTQP,
                                  const TensorQuantizationParams &inputTQP,
                                  const TensorQuantizationParams &weightsTQP,
-                                 Schema schema, ElemKind biasQTy) {
+                                 Schema schema, ElemKind biasQTy,
+                                 bool biasZero) {
   // Choose bias offset. For int32 bias we always force offset 0 in order
   // to simplify the implementation since the dynamic range allows it.
   int32_t biasOffset = biasTQP.offset;
   if (biasQTy == ElemKind::Int32QTy) {
     biasOffset = 0;
   }
   // Choose bias scale. We try to force the bias scale value to the product
-  // scaleInput * scaleWeights but only if the resulting scale is larger
+  // inputScale * weightsScale but only if the resulting scale is larger
   // (in order to avoid bias data saturation).
-  float scaleInput = inputTQP.scale;
-  float scaleWeights = weightsTQP.scale;
+  float inputScale = inputTQP.scale;
+  float weightsScale = weightsTQP.scale;
   float biasScale = biasTQP.scale;
-  if (scaleInput * scaleWeights > biasScale) {
-    biasScale = scaleInput * scaleWeights;
+  if (inputScale * weightsScale >= biasScale || biasZero) {
+    biasScale = inputScale * weightsScale;
   }
   // Validate new bias TQP and return.
   TensorQuantizationParams biasTQPNew = {biasScale, biasOffset};
   validateQuantizationParams(biasTQPNew, schema, biasQTy);
   return biasTQPNew;
 }
 
+void specializeBiasWeightsQuantizationParams(
+    TensorQuantizationParams &biasTQP, const TensorQuantizationParams &inputTQP,
+    TensorQuantizationParams &weightsTQP, Schema schema, ElemKind biasQTy,
+    bool biasZero) {
+  // Choose bias offset. For int32 bias we always force offset 0 in order
+  // to simplify the implementation since the dynamic range allows it.
+  if (biasQTy == ElemKind::Int32QTy) {
+    biasTQP.offset = 0;
+  }
+  // Choose bias scale. We try to force the bias scale value to the product
+  // inputScale * weightsScale but only if the resulting scale is larger
+  // (in order to avoid bias data saturation). Otherwise, for INT32 bias
+  // only, we change the weightsScale to enforce the equality.
+  float inputScale = inputTQP.scale;
+  float weightsScale = weightsTQP.scale;
+  float biasScale = biasTQP.scale;
+  if (inputScale * weightsScale >= biasScale || biasZero) {
+    biasScale = inputScale * weightsScale;
+  } else {
+    if (biasQTy == ElemKind::Int32QTy) {
+      weightsScale = biasScale / inputScale;
+      biasScale = inputScale * weightsScale;
```

Comment:
@842974287 In float32 arithmetic when computing `a = b / c` this does not mean that afterwards `b == a * c` (bit-exactly) because float32 division looses some precision. Therefore the second statement is forcing explicitly that `biasScale == inputScale * weightsScale` (bit-exactly). This equality is required in order to guarantee that at runtime for int32 bias we use shifts of 0 and a scale of 1 (which is easier to compute).